ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs

```diff
@@ -203,13 +237,58 @@ prop_simple_cardano_convergence TestSetup
     epochSizeShelley :: EpochSize
     epochSizeShelley = sgEpochLength genesisShelley
 
+    -- the protocol version of the Byron era proposal
+    propPV :: CC.Update.ProtocolVersion
+    propPV =
+      if setupHardFork
+      then CC.Update.ProtocolVersion shelleyMajorVersion 0 0
+      else CC.Update.ProtocolVersion
+             byronMajorVersion
+             (byronInitialMinorVersion + 1)
+             0
+
+    reachesShelley :: ReachesShelley
+    reachesShelley = ReachesShelley
+      { rsByronSlots =
+          BoolProps.enabledIf $ t > numByronEpochs * unEpochSize epochSizeByron
+      , rsPV         = BoolProps.enabledIf setupHardFork
+      , rsShelleyBlocks =
+          or $
+          [ isShelley blk
+          | (_nid, no) <- Map.toList testOutputNodes
+          , let NodeOutput{nodeOutputFinalChain} = no
+          , blk <- chainToList nodeOutputFinalChain
+          ]
+      , rsShelleySlots =
+          assert (w >= k) $
+          BoolProps.requiredIf $ t > w + 1 - k
+            -- logic: if we are ensured at least @k@ blocks in @w@ slots, then
+            -- we're ensured at least @1@ block in @w+1-k@ slots
+      }
+      where
+        TestConfig{numSlots} = setupTestConfig
+        NumSlots t           = numSlots
+        TestOutput{testOutputNodes} = testOutput
+
+        k :: Word64
+        k = maxRollbacks setupK
+
+        w :: Word64
+        w = Shelley.computeStabilityWindow
+            setupK
+            (SL.sgActiveSlotCoeff genesisShelley)
+
+        isShelley :: HF.HardForkBlock '[ByronBlock, ShelleyBlock c] -> Bool
+        isShelley (HF.HardForkBlock (HF.OneEraBlock ns)) = index_NS ns == 1
```

Comment:
You can use the `BlockShelley` pattern synonym from `O.C.Cardano.Block`, that's much prettier:
```haskell
isShelley :: CardanoBlock scc -> Bool
isShelley (BlockShelley _) = True
isShelley _                = False
```

Reply:
Done! Thanks, much nicer.